rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,302 @@
+state: draft
+---
+
+# RFD 0144 - Client Tools Updates
+
+## Required Approvers
+
+* Engineering: @sclevine && @bernardjkim && @r0mant
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What/Why
+
+This RFD describes how client tools like `tsh` and `tctl` can be kept up to
+date, either using automatic updates or self-managed updates.
+
+Keeping client tools updated helps with security (fixes for known security
+vulnerabilities are pushed to endpoints), bugs (fixes for resolved issues are
+pushed to endpoints), and compatibility (users no longer have to learn and
+understand [Teleport component
+compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility)
+rules).
+
+## Details
+
+### Summary
+
+Client tools like `tsh` and `tctl` will automatically download and install the
+required version for the Teleport cluster.
+
+Enrollment in automatic updates for client tools will be controlled at the
+cluster level. By default all clusters will be opted into automatic updates for
+client tools. Cluster administrators using MDM software like JamF will be able
+opt-out manually manage updates.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
+
+#### Client tools
+
+##### Automatic updates
+
+When `tsh login` is executed, client tools will check `/v1/webapi/ping` to
+determine if automatic updates are enabled. If the cluster's required version
+differs from the current binary, client tools will download and re-execute
+using the version required by the cluster.
+
+The original client tools binaries won't be overwritten. Instead, additional
+binaries will be downloaded and stored in `~/.tsh/versions` with `0755`
+permissions.  A file named `versions.yaml` in the same directory will maintain
+the mapping between proxies and tools versions, to support auto-purge and free
+up disk space from unused binaries.
+
+To enable concurrent operation of client tools, a locking mechanisms utilizing
+[syscall.Flock](https://pkg.go.dev/syscall#Flock) (for Linux and macOS) and
+[LockFileEx](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex)
+(for Windows) will be used.
+
+A side effect of automatic updates will be users will be able to connect to
+different Teleport clusters (running different versions of Teleport, including
+pre-releases) without juggling multiple versions of client tools.
+
+```
+$ tree ~/.tsh
+~/.tsh
+├── current-profile
+├── keys
+│  └── proxy.example.com
+│     ├── cas
+│     │  └── example.com.pem
+│     ├── certs.pem
+│     ├── foo
+│     ├── foo-ssh
+│     │  └── example.com-cert.pub
+│     ├── foo-x509.pem
+│     └── foo.pub
+├── known_hosts
+├── proxy.example.com.yaml
+└── versions
+   ├── 15.0.0
+   │  └── bin
+   │     ├── tctl
+   │     └── tsh
+   ├── 15.1.1
+   │  └── bin
+   │     ├── tctl
+   │     └── tsh
+   └── versions.yaml
+```
+
+Users can cancel client tools updates using `Ctrl-C`. This may be needed if the
```

Comment:
With this approach, we'll always download and re-execute from a `.tar.gz` archive.

That should actually resolve this issue on macOS right? No matter the install method (let's say you start with a unsigned package), the binary you execute will be a signed binary that can be used with Touch ID.

Reply:
Quick reply, will read the complete RFD later:

Touch ID needs the ".app" shell to work, the signed binary by itself doesn't carry all that is required.

Keeping the spirit of the design, you'll likely want something like this:

```
~/.tsh
└── versions
   ├── 15.0.0
   │  └── bin
   │     ├── tctl -> link to an eventual tctl.app?
   │     └── tsh  -> ../tsh.app/Contents/MacOS/tsh
   │  └── tsh.app
```

Also do note that macOS keeps tabs of where ".apps" are, and it can get wonky if you have multiple versions of the same app (as it may pick an unexpected one to execute on say, opened links). This is why we make [non-relocatable bundles][1]. Similarly for Connect, if that is a question.

[1]: https://github.com/gravitational/teleport/blob/d2dd5d395d030e76846781b6a5f79ac164dcbd77/build.assets/build-pkg-tsh.sh#L26

Reply:
I tried using `pkgutil` and it appears to work, but I will do more testing.

```
$ curl -LO https://cdn.teleport.dev/tsh-15.2.0.pkg 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 66.3M  100 66.3M    0     0  10.9M      0  0:00:06  0:00:06 --:--:-- 11.1M
$ pkgutil --expand-full tsh-15.2.0.pkg tsh-15-pkg/
$ ./tsh-15-pkg/Payload/tsh.app/Contents/MacOS/tsh touchid diag
Has compile support? true
Has signature? true
Has entitlements? true
Passed LAPolicy test? true
Passed Secure Enclave test? true
Touch ID enabled? true
$
$ curl -LO https://cdn.teleport.dev/tsh-14.3.13.pkg
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 74.8M  100 74.8M    0     0  10.0M      0  0:00:07  0:00:07 --:--:-- 11.1M
$ pkgutil --expand-full tsh-14.3.13.pkg tsh-14-pkg/
$ ./tsh-14-pkg/Payload/tsh.app/Contents/MacOS/tsh touchid diag
Has compile support? true
Has signature? true
Has entitlements? true
Passed LAPolicy test? true
Passed Secure Enclave test? true
Touch ID enabled? true
```

Reply:
`pkgutil` should work fine, although I can't speak for the particular invocation. I _think_ pkgs are just some kind of zip, so it may be possible to forego the external binary if you want.

Reply:
I think it’s xar file and we can use `xz` to unpack it 🥲

Reply:
@codingllama Any issue with the above approach of extracting the application bundle and re-executing it? Seems to work?